scripts/import_dvv.sql

```diff
@@ -197,20 +372,47 @@ select distinct on ("huoneiston vanhin asukas (henkilötunnus)")
 from jkr_dvv.vanhin
 where
     vanhin."huoneiston vanhin asukas (henkilötunnus)" is not null
-on conflict do nothing; -- some elders (e.g. owners) already exist
-
+on conflict (henkilotunnus) do update -- some elders (e.g. owners) already exist
+-- updates existing osapuoli
+set
+    nimi = excluded.nimi,
+    katuosoite = excluded.katuosoite,
+    postitoimipaikka = excluded.postitoimipaikka,
+    postinumero = excluded.postinumero,
+    kunta = excluded.kunta,
+    tiedontuottaja_tunnus = excluded.tiedontuottaja_tunnus
+where
+    jkr.osapuoli.henkilotunnus = excluded.henkilotunnus;
+    
 -- Insert elders to jkr.rakennuksen_vanhimmat
-insert into jkr.rakennuksen_vanhimmat (rakennus_id, osapuoli_id, huoneistokirjain, huoneistonumero, jakokirjain, alkupvm)
+insert into jkr.rakennuksen_vanhimmat (rakennus_id, osapuoli_id, huoneistokirjain, huoneistonumero, jakokirjain, alkupvm, loppupvm)
 select
     (select id from jkr.rakennus where vanhin.rakennustunnus = rakennus.prt) as rakennus_id,
     (select id from jkr.osapuoli where vanhin."huoneiston vanhin asukas (henkilötunnus)" = osapuoli.henkilotunnus and osapuoli.tiedontuottaja_tunnus = 'dvv') as osapuoli_id,
     nullif(vanhin."huo_neisto_kirjain", ' ') as huoneistokirjain,
     nullif(vanhin."huo_neisto_numero", '000')::integer as huoneistonumero,
     nullif(vanhin."jako_kirjain", ' ') as jakokirjain,
-    to_date(vanhin."vakin kotim osoitteen alkupäivä"::text, 'YYYYMMDD') as alkupvm
-    -- TODO! Insert loppupvm when updating with new dvv-material.
+    to_date(vanhin."vakin kotim osoitteen alkupäivä"::text, 'YYYYMMDD') as alkupvm,
+    null as loppupvm
 from jkr_dvv.vanhin
 where
     vanhin."huoneiston vanhin asukas (henkilötunnus)" is not null and
     exists (select 1 from jkr.rakennus where vanhin.rakennustunnus = rakennus.prt) -- not all buildings are listed
-on conflict do nothing;
+-- maybe improve so this only updates elders when atleast one value has changed.
+-- on conflict (building_id, osapuoli_id) do update -- No constraint for building_id and osapuoli_id.
+on conflict (osapuoli_id) do update 
```

Comment:
Tajusin juuri että tämä vaatii kaiketi myös tietokantamigraation :)

Eli tällä hetkellä tietokannassa osapuoli voi olla vanhin asukas vain yhdessä huoneistossa, rakennuksen_vanhimmat-taulun primary key on pelkkä osapuoli_id.

Tämä pitäisi muuttaa niin, että sama henkilö voi asua eri aikoina eri huoneistoissa ja rakennuksissa. Eli käytännössä primary keyn pitäisi olla sitten varmaankin kaikki viisi saraketta
(rakennus_id, osapuoli_id, huoneistokirjain, huoneistonumero, jakokirjain), mutta silloinkin ongelma olisi tapaus, jossa sama henkilö muuttaa takaisin asuntoon jossa on asunut aiemmin.

Eli hmm, nyt alkaa näyttää siltä että pitää tehdä rakennuksen_vanhimmat-tauluun ihan oma id-primary key, jotta näitä vanhimpia voi lisätä. Eli sen jälkeen tätä konfliktia ei enää tule, vaan päivämäärien päivitys pitää hoitaa vain tuolla triggerillä.

Muita kenttiä ei päivitetä, vaan aina jos osapuoli muuttaa vaikka samassa talossa eri huoneistoon, vanhaan vanhimmat-riviin lisätään loppupäivämäärä, ja sitten uuteen vanhimmat-riviin (jossa eri huoneistonumero) tulee vain alkupäivämäärä, ja null-loppupäivämäärä.

Reply:
Ja samasta teemasta: koska nyt omistajuudella on myös alku- ja loppupäivämäärä, tajusin että myös rakennuksen_omistajat-tauluun pitänee lisätä id ja poistaa vanha primary key (rakennus_id, osapuoli_id). Tämä sen takia, että sama henkilö voi myydä osuutensa rakennuksesta, ja seuraavana vuonna vaikka ostaa sen takaisin itselleen. Eli omistaja voi kadota yhtenä vuonna ja sen jälkeen ilmestyä takaisin.